library/cartridge_check_instance_started.py

```diff
@@ -4,32 +4,91 @@
 from ansible.module_utils.helpers import ModuleRes, CartridgeException
 from ansible.module_utils.helpers import get_control_console
 
-
 argument_spec = {
     'control_sock': {'required': True, 'type': 'str'},
     'stateboard': {'required': True, 'type': 'bool'},
+    'expected_states': {'required': False, 'type': 'list', 'default': ['Unconfigured', 'RolesConfigured']},
+    'check_buckets_are_discovered': {'required': False, 'type': 'bool', 'default': True},
 }
 
 
-def check_instance_started(params):
-    try:
-        control_console = get_control_console(params['control_sock'])
-        instance_is_alive = True
-        if not params['stateboard']:
-            instance_is_alive, _ = control_console.eval_res_err('''
-                return require('membership').myself().status == 'alive'
-            ''')
-    except CartridgeException as e:
-        return ModuleRes(success=False, msg=str(e))
+def check_stateboard_started(control_console):
+    box_status, err = control_console.eval_res_err('''
+        if type(box.cfg) == 'function' or box.cfg.listen == nil then
+            return nil, "box hasn't been configured"
+        end
+        return true
+    ''')
+    if not box_status:
+        return ModuleRes(success=False, msg="Stateboard is not running: %s" % err)
+
+    return ModuleRes(success=True)
+
+
+def check_instance_started(control_console, expected_states, check_buckets_are_discovered):
+    instance_state, err = control_console.eval_res_err('''
+        return require('cartridge.confapplier').get_state()
+    ''')
+    if not instance_state:
+        return ModuleRes(success=False, msg="Impossible to get state: %s" % err)
+    if instance_state not in expected_states:
+        return ModuleRes(
+            success=False,
+            msg="Instance is not in one of states: %s, it's in '%s' state" % (
+                expected_states,
+                instance_state,
+            ),
+        )
+
+    if check_buckets_are_discovered:
+        buckets_ok, err = control_console.eval_res_err('''
+            local vshard_utils = require('cartridge.vshard-utils')
+            local vshard_router = require('cartridge.roles.vshard-router')
 
-    return ModuleRes(success=instance_is_alive)
+            local groups = vshard_utils.get_known_groups()
+            for group_name, group_opts in pairs(groups) do
+                if group_opts.bootstrapped == true then
+                    local router = vshard_router.get(group_name)
```

Comment:
Seems that this part can be moved to some `check_group_buckets` function. Just to decrease number of nesting levels.